StaticStrings: I'm trying to add an iterate method , how can I fix an undefined reference to `ijl_trhow`?

Hi there,
I'm starting to play with StaticTools.jl, and it's great. However, I'd like to use methods like `startswith` (I'm trying to statically compile my package Norg.jl). I decided to try an implement the required `iterate` method, by basically translating [the standard library](https://github.com/JuliaLang/julia/blob/master/base/strings/string.jl#L207-L235) as follow:

```julia
using StaticTools, StaticCompiler
function Base.iterate(s::StaticString, i::Int=firstindex(s))
       ((i % UInt) - 1 < ncodeunits(s) && s[i] ≠ 0x00) || return nothing
           b = @inbounds codeunit(s, i)
           u = UInt32(b) << 24
           between(b, 0x80, 0xf7) || return reinterpret(Char, u), i+1
           return iterate_continued(s, i, u)
           end
@inline between(b::T, lo::T, hi::T) where {T<:Integer} = (lo ≤ b) & (b ≤ hi)
function iterate_continued(s::StaticString, i::Int, u::UInt32)
           u < 0xc0000000 && (i += 1; @goto ret)
           n = ncodeunits(s)
           # first continuation byte
           (i += 1) > n && @goto ret
           @inbounds b = codeunit(s, i)
           b & 0xc0 == 0x80 || @goto ret
           u |= UInt32(b) << 16
           # second continuation byte
           ((i += 1) > n) | (u < 0xe0000000) && @goto ret
           @inbounds b = codeunit(s, i)
           b & 0xc0 == 0x80 || @goto ret
           u |= UInt32(b) << 8
           # third continuation byte
           ((i += 1) > n) | (u < 0xf0000000) && @goto ret
           @inbounds b = codeunit(s, i)
           b & 0xc0 == 0x80 || @goto ret
           u |= UInt32(b); i += 1
       @label ret
           return reinterpret(Char, u), i
       end
```

And this works fine in the interpreter

```julia
julia> startswith(c"foobar", c"foo")
true
```

But this won't compile:

```julia
julia> function test()
           if startswith(c"foobar", c"foo")
           println(c"marche")
           else
           println(c"marche pas")
           end
           end
test (generic function with 1 method)

julia> compile_executable(test, (), "./", filename="test_compile")
/usr/bin/x86_64-linux-gnu-ld : ./test_compile.o : dans la fonction « gpu_gc_pool_alloc » :
text:(.text+0x420) : référence indéfinie vers « ijl_throw »
clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
```

I guess there is some exception throw somewhere in the code but I have no clue as to how I can find them. Do you have any insight on that?

So this will still not compile (though ultimately because of the `_gpu_gc_pool_alloc`, not the `_ijl_throw`):
```
julia> compile_executable(test, (), "./", filename="test_compile")
Undefined symbols for architecture arm64:
  "_ijl_throw", referenced from:
      _gpu_gc_pool_alloc in test_compile.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ERROR: failed process: Process(`cc -e _julia_test ./test_compile.o -o ./test_compile`, ProcessExited(1)) [1]

Stacktrace:
 [1] pipeline_error
   @ ./process.jl:565 [inlined]
 [2] run(::Cmd; wait::Bool)
   @ Base ./process.jl:480
 [3] run
   @ ./process.jl:477 [inlined]
 [4] generate_executable(f::Function, tt::Type, path::String, name::String, filename::String; cflags::Cmd, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ StaticCompiler ~/.julia/packages/StaticCompiler/clvAO/src/StaticCompiler.jl:375
 [5] compile_executable(f::Function, types::Tuple{}, path::String, name::String; filename::String, cflags::Cmd, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ StaticCompiler ~/.julia/packages/StaticCompiler/clvAO/src/StaticCompiler.jl:244
 [6] top-level scope
   @ REPL[31]:1
```

--- a/src/abstractstaticstring.jl
+++ b/src/abstractstaticstring.jl
@@ -122,3 +122,66 @@
         end
         return false
     end
+    
+    # Adapted from Julia's stdlib
+    """
+        iterate(s::AbstractStaticString, i=firstindex(s))
+
+    Adapted form Julia's stdlib, but made type-stable. 
+
+    !!! warning "Return type"
+        
+        The interface is a bit different from `Base`. When iterating outside of
+        the string, it will return the Null character and the current index.
+
+    # Examples
+
+    ```jldoctest
+    julia> s = c"foo"
+    c"foo"
+
+    julia> iterate(s, 1)
+    ('f', 2)
+
+    julia> iterate(s, 99999)
+    ('\\0', 99999)
+    ```
+    """
+    @inline function Base.iterate(s::AbstractStaticString, i::Int=firstindex(s))
+        ((i % UInt) - 1 < ncodeunits(s) && s[i] ≠ 0x00) || return ('\0', i)
+        b = @inbounds codeunit(s, i)
+        u = UInt32(b) << 24
+        between(b, 0x80, 0xf7) || return reinterpret(Char, u), i+1
+        return iterate_continued(s, i, u)
+    end
+    @inline between(b::T, lo::T, hi::T) where {T<:Integer} = (lo ≤ b) & (b ≤ hi)
+    @inline function iterate_continued(s::AbstractStaticString, i::Int, u::UInt32)
+        u < 0xc0000000 && (i += 1; @goto ret)
+        n = ncodeunits(s)
+        # first continuation byte
+        (i += 1) > n && @goto ret
+        @inbounds b = codeunit(s, i)
+        b & 0xc0 == 0x80 || @goto ret
+        u |= UInt32(b) << 16
+        # second continuation byte
+        ((i += 1) > n) | (u < 0xe0000000) && @goto ret
+        @inbounds b = codeunit(s, i)
+        b & 0xc0 == 0x80 || @goto ret
+        u |= UInt32(b) << 8
+        # third continuation byte
+        ((i += 1) > n) | (u < 0xf0000000) && @goto ret
+        @inbounds b = codeunit(s, i)
+        b & 0xc0 == 0x80 || @goto ret
+        u |= UInt32(b); i += 1
+        @label ret
+        return reinterpret(Char, u), i
+    end
+    @inline function Base.startswith(a::AbstractStaticString, b::AbstractStaticString)
+           i, j = iterate(a), iterate(b)
+           while true
+               j[1] === '\0' && return true # ran out of prefix: success!
+               i[1] === '\0' && return false # ran out of source: failure
+               i[1] == j[1] || return false # mismatch: failure
+               i, j = iterate(a, i[2]), iterate(b, j[2])
+           end
+       end

--- a/src/staticstring.jl
+++ b/src/staticstring.jl
@@ -144,3 +144,4 @@
     # Implement some of the AbstractString interface -- where overriding AbstractStaticString defaults
     @inline Base.ncodeunits(s::StaticString{N}) where N = N
     @inline Base.codeunits(s::StaticString) = s.data
+

--- a/test/teststaticstring.jl
+++ b/test/teststaticstring.jl
@@ -62,3 +62,7 @@
     @test contains(str, c"foo")
     @test contains(str, c"bar")
     @test contains(str, c"baz")
+    @test startswith(str, c"foo")
+    @test !startswith(c"foo", str)
+    @test !startswith(str, c"g")
+

